File: backend/src/secrets/secretsProvider.js
/**
 * Secrets-provider abstraction — enterprise pattern for loading credentials outside Git.
 *
 * Production teams point SECRETS_PROVIDER=aws at real AWS Secrets Manager.
 * Local Docker uses SECRETS_PROVIDER=mock-aws and the mock-secrets-manager container.
 * Unit tests and CI use SECRETS_PROVIDER=file with committed ci.secrets (fake values only).
 */
const fs = require("fs");
const path = require("path");

/** Default mock AWS Secrets Manager base URL inside Docker Compose. */
const DEFAULT_MOCK_AWS_URL = "http://mock-secrets-manager:4566";

/**
 * Resolve which on-disk secrets bundle matches the deployment slice (dev/staging/prod/ci).
 * @param {string} deploymentEnv
 * @returns {string} absolute path to a *.secrets file
 */
function resolveSecretsFilePath(deploymentEnv) {
  if (process.env.SECRETS_FILE) {
    return path.resolve(process.env.SECRETS_FILE);
  }
  const slice = String(deploymentEnv || "dev").toLowerCase();
  return path.resolve(__dirname, `../../${slice}.secrets`);
}

/**
 * Parse dotenv-style KEY=VALUE lines into a plain object (no shell expansion).
 * @param {string} raw
 * @returns {Record<string, string>}
 */
function parseSecretsText(raw) {
  const out = {};
  String(raw || "")
    .split(/\r?\n/)
    .forEach((line) => {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) {
        return;
      }
      const eq = trimmed.indexOf("=");
      if (eq <= 0) {
        return;
      }
      const key = trimmed.slice(0, eq).trim();
      const value = trimmed.slice(eq + 1).trim();
      if (key) {
        out[key] = value;
      }
    });
  return out;
}

/**
 * Read secrets directly from a gitignored *.secrets file (fallback when mock AWS is down).
 * @param {string} filePath
 * @returns {Record<string, string>}
 */
function loadSecretsFromFile(filePath) {
  if (!fs.existsSync(filePath)) {
    return {};
  }
  const raw = fs.readFileSync(filePath, "utf8");
  return parseSecretsText(raw);
}

/**
 * Build the AWS-style secret id, e.g. triage/dev — overridable via SECRETS_BUNDLE_ID.
 * @param {string} deploymentEnv
 * @returns {string}
 */
function secretBundleId(deploymentEnv) {
  return process.env.SECRETS_BUNDLE_ID || `triage/${String(deploymentEnv || "dev").toLowerCase()}`;
}

/**
 * Fetch one secret bundle from the mock AWS Secrets Manager HTTP API (GetSecretValue shape).
 * @param {string} deploymentEnv
 * @returns {Promise<Record<string, string>>}
 */
async function loadSecretsFromMockAws(deploymentEnv) {
  const baseUrl = (process.env.SECRETS_MANAGER_URL || DEFAULT_MOCK_AWS_URL).replace(/\/$/, "");
  const secretId = secretBundleId(deploymentEnv);
  const url = `${baseUrl}/v1/secrets/${encodeURIComponent(secretId)}`;

  let response;
  try {
    response = await fetch(url, { method: "GET" });
  } catch (err) {
    throw new Error(`mock AWS Secrets Manager unreachable at ${url}: ${err.message}`);
  }

  if (!response.ok) {
    throw new Error(
      `mock AWS Secrets Manager returned HTTP ${response.status} for secretId=${secretId}`
    );
  }

  const body = await response.json();
  const secretString = body.SecretString || "";
  if (typeof secretString === "string" && secretString.trim().startsWith("{")) {
    try {
      return JSON.parse(secretString);
    } catch {
      return parseSecretsText(secretString);
    }
  }
  return parseSecretsText(String(secretString));
}

/**
 * Create the active secrets provider based on SECRETS_PROVIDER (mock-aws | file | aws).
 * @param {{ deploymentEnv?: string }} [options]
 * @returns {{ load: () => Promise<Record<string, string>>, providerName: string }}
 */
function createSecretsProvider(options = {}) {
  const deploymentEnv = (
    options.deploymentEnv ||
    process.env.DEPLOYMENT_ENV ||
    process.env.APP_ENV ||
    "dev"
  ).toLowerCase();
  const providerName = (process.env.SECRETS_PROVIDER || "mock-aws").toLowerCase();

  if (providerName === "file") {
    const filePath = resolveSecretsFilePath(deploymentEnv);
    return {
      providerName: "file",
      load: async () => loadSecretsFromFile(filePath),
    };
  }

  if (providerName === "mock-aws" || providerName === "aws") {
    return {
      providerName,
      load: async () => {
        try {
          return await loadSecretsFromMockAws(deploymentEnv);
        } catch (err) {
          // Graceful fallback: read gitignored *.secrets when mock service is not running (local npm test).
          const filePath = resolveSecretsFilePath(deploymentEnv);
          const fromFile = loadSecretsFromFile(filePath);
          if (Object.keys(fromFile).length > 0) {
            return fromFile;
          }
          throw err;
        }
      },
    };
  }

  throw new Error(`Unsupported SECRETS_PROVIDER=${providerName}`);
}

/**
 * Merge secret key/value pairs into process.env without overwriting existing shell exports.
 * @param {Record<string, string>} secrets
 * @param {{ override?: boolean }} [options]
 */
function applySecretsToProcessEnv(secrets, options = {}) {
  const override = options.override !== false;
  Object.entries(secrets || {}).forEach(([key, value]) => {
    if (override || process.env[key] === undefined) {
      process.env[key] = String(value);
    }
  });
}

module.exports = {
  DEFAULT_MOCK_AWS_URL,
  resolveSecretsFilePath,
  parseSecretsText,
  loadSecretsFromFile,
  loadSecretsFromMockAws,
  createSecretsProvider,
  applySecretsToProcessEnv,
  secretBundleId,
};

File: ai_service/app/secrets_provider.py
"""
Python secrets-provider — mirrors Node secretsProvider.js for Celery/Kafka workers.

Pattern: SECRETS_PROVIDER=mock-aws (dev Docker) or aws (staging/prod IAM) at process start.
Technology: urllib mock HTTP for dev; boto3 GetSecretValue for real AWS Secrets Manager.
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Dict

# Repository root when running in Docker (/app) or locally (parent of ai_service).
REPO_ROOT = Path(__file__).resolve().parents[2]
BACKEND_DIR = REPO_ROOT / "backend"


def _deployment_env() -> str:
    """Return dev/staging/prod/ci slice from DEPLOYMENT_ENV or APP_ENV."""
    return (os.getenv("DEPLOYMENT_ENV") or os.getenv("APP_ENV") or "dev").lower()


def _secrets_file_path() -> Path:
    """Resolve gitignored *.secrets path for the active deployment slice."""
    explicit = os.getenv("SECRETS_FILE")
    if explicit:
        return Path(explicit).resolve()
    return BACKEND_DIR / f"{_deployment_env()}.secrets"


def parse_secrets_text(raw: str) -> Dict[str, str]:
    """Parse dotenv-style KEY=VALUE lines (no shell expansion)."""
    out: Dict[str, str] = {}
    for line in raw.splitlines():
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        if "=" not in trimmed:
            continue
        key, value = trimmed.split("=", 1)
        key = key.strip()
        if key:
            out[key] = value.strip()
    return out


def load_secrets_from_file(path: Path | None = None) -> Dict[str, str]:
    """Read secrets directly from a gitignored *.secrets file."""
    target = path or _secrets_file_path()
    if not target.is_file():
        return {}
    return parse_secrets_text(target.read_text(encoding="utf-8"))


def secret_bundle_id() -> str:
    """AWS-style secret name, e.g. triage/dev."""
    return os.getenv("SECRETS_BUNDLE_ID") or f"triage/{_deployment_env()}"


def load_secrets_from_mock_aws() -> Dict[str, str]:
    """HTTP GET to mock AWS Secrets Manager (GetSecretValue-compatible JSON)."""
    base = (os.getenv("SECRETS_MANAGER_URL") or "http://mock-secrets-manager:4566").rstrip("/")
    secret_id = secret_bundle_id()
    url = f"{base}/v1/secrets/{urllib.parse.quote(secret_id, safe='')}"

    try:
        with urllib.request.urlopen(url, timeout=5) as response:
            body = json.loads(response.read().decode("utf-8"))
    except (urllib.error.URLError, TimeoutError) as err:
        raise RuntimeError(f"mock AWS Secrets Manager unreachable at {url}: {err}") from err

    secret_string = body.get("SecretString") or ""
    if isinstance(secret_string, str) and secret_string.strip().startswith("{"):
        try:
            parsed = json.loads(secret_string)
            return {str(k): str(v) for k, v in parsed.items()}
        except json.JSONDecodeError:
            pass
    return parse_secrets_text(str(secret_string))


def load_secrets_from_aws() -> Dict[str, str]:
    """Fetch secret bundle from real AWS Secrets Manager (boto3 + IAM role or env keys)."""
    import boto3
    from botocore.exceptions import ClientError

    region = os.getenv("AWS_REGION") or os.getenv("AWS_DEFAULT_REGION") or "us-east-1"
    secret_id = secret_bundle_id()
    client = boto3.client("secretsmanager", region_name=region)
    try:
        response = client.get_secret_value(SecretId=secret_id)
    except ClientError as err:
        raise RuntimeError(
            "AWS Secrets Manager get_secret_value failed for "
            f"secretId={secret_id} region={region}: {err}"
        ) from err

    secret_string = response.get("SecretString") or ""
    if isinstance(secret_string, str) and secret_string.strip().startswith("{"):
        try:
            parsed = json.loads(secret_string)
            return {str(k): str(v) for k, v in parsed.items()}
        except json.JSONDecodeError:
            pass
    return parse_secrets_text(str(secret_string))


def load_application_secrets() -> Dict[str, str]:
    """Load secrets using SECRETS_PROVIDER and merge into os.environ."""
    provider = (os.getenv("SECRETS_PROVIDER") or "mock-aws").lower()

    if provider == "file":
        secrets = load_secrets_from_file()
    elif provider == "mock-aws":
        try:
            secrets = load_secrets_from_mock_aws()
        except RuntimeError:
            secrets = load_secrets_from_file()
            if not secrets:
                raise
    elif provider == "aws":
        secrets = load_secrets_from_aws()
    else:
        raise ValueError(f"Unsupported SECRETS_PROVIDER={provider}")

    for key, value in secrets.items():
        os.environ[key] = value
    return secrets


def inject_secrets_into_environ() -> str:
    """
    Shell-export lines for docker-entrypoint-with-secrets.sh (eval-friendly).
    Returns empty string when no secrets loaded.
    """
    secrets = load_application_secrets()
    lines = []
    for key, value in secrets.items():
        # Single-quote escape for bash eval safety.
        safe = value.replace("'", "'\\''")
        lines.append(f"export {key}='{safe}'")
    return "\n".join(lines)
